src/Serilog.Sinks.PeriodicBatching/Sinks/PeriodicBatching/PeriodicBatchingSink.cs

```diff
@@ -56,305 +58,209 @@ public class PeriodicBatchingSink : ILogEventSink, IDisposable, IBatchedLogEvent
+            
+            var fillBatch = Task.Delay(_batchScheduler.NextInterval);
             do
             {
-                while (_waitingBatch.Count < _batchSizeLimit &&
-                       _queue.TryDequeue(out var next))
+                while (_currentBatch.Count < _batchSizeLimit &&
+                       !_shutdownSignal.IsCancellationRequested &&
+                       _queue.Reader.TryRead(out var next))
                 {
-                    _waitingBatch.Enqueue(next);
+                    _currentBatch.Enqueue(next);
                 }
+            } while ((_currentBatch.Count < _batchSizeLimit || _currentBatch.Count > 0 && isEagerBatch) &&
+                     !_shutdownSignal.IsCancellationRequested &&
+                     await TryWaitToReadAsync(_queue.Reader, fillBatch, _shutdownSignal.Token).ConfigureAwait(false));
 
-                if (_waitingBatch.Count == 0)
+            try
+            {
+                if (_currentBatch.Count == 0)
                 {
-                    await _batchedLogEventSink.OnEmptyBatchAsync().ConfigureAwait(false);
-                    return;
+                    await _targetSink.OnEmptyBatchAsync().ConfigureAwait(false);
+                }
+                else
+                {
+                    isEagerBatch = false;
+                    await _targetSink.EmitBatchAsync(_currentBatch).ConfigureAwait(false);
                 }
 
-                await _batchedLogEventSink.EmitBatchAsync(_waitingBatch).ConfigureAwait(false);
-
-                batchWasFull = _waitingBatch.Count >= _batchSizeLimit;
-                _waitingBatch.Clear();
-                _status.MarkSuccess();
+                _currentBatch.Clear();
+                _batchScheduler.MarkSuccess();
             }
-            while (batchWasFull); // Otherwise, allow the period to elapse
-        }
-        catch (Exception ex)
-        {
-            SelfLog.WriteLine("Exception while emitting periodic batch from {0}: {1}", this, ex);
-            _status.MarkFailure();
-        }
-        finally
-        {
-            if (_status.ShouldDropBatch)
-                _waitingBatch.Clear();
-
-            if (_status.ShouldDropQueue)
+            catch (Exception ex)
             {
-                while (_queue.TryDequeue(out _)) { }
-            }
+                SelfLog.WriteLine($"PeriodicBatchingSink ({_targetSink}) failed emitting a batch: {ex}");
+                _batchScheduler.MarkFailure();
+                
+                if (_batchScheduler.ShouldDropBatch)
+                    _currentBatch.Clear();
 
-            lock (_stateLock)
-            {
-                if (!_unloading)
-                    SetTimer(_status.NextInterval);
+                if (_batchScheduler.ShouldDropQueue)
+                {
+                    // This is not ideal; the goal is to reduce memory pressure on the client if the server is offline
```

Comment:
this should at least selflog (maybe also when buffer is full and TryWrite is refused?)